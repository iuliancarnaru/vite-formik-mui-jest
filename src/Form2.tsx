import { Form, Formik, FormikProps } from 'formik';
import * as yup from 'yup';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';

const schema = yup.object({
  email: yup
    .string()
    .email('Enter a valid email')
    .required('Email is required'),
  password: yup
    .string()
    .min(8, 'Password should be of minimum 8 characters length')
    .required('Password is required'),
});

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

interface MyFormValues {
  email: string;
  password: string;
}

interface MaterialUIForm2Props {
  onSubmit: (values: MyFormValues) => void;
}

export const MaterialUIForm2 = ({ onSubmit }: MaterialUIForm2Props) => {
  const initialValues: MyFormValues = {
    email: '',
    password: '',
  };

  const handleSubmit = async ({ email, password }: MyFormValues) => {
    await sleep(500);
    onSubmit({ email, password });
  };

  return (
    <div style={{ maxWidth: '300px', padding: '20px' }}>
      <Formik
        initialValues={initialValues}
        validationSchema={schema}
        onSubmit={(values, { setSubmitting, resetForm }) => {
          handleSubmit(values);

          setTimeout(() => {
            resetForm();
            setSubmitting(false);
          }, 2000);
        }}
        validateOnMount
      >
        {({
          values,
          handleChange,
          touched,
          errors,
          isSubmitting,
          isValid,
        }: FormikProps<MyFormValues>) => (
          <Form>
            <TextField
              fullWidth
              id="email"
              name="email"
              label="Email"
              type="email"
              variant="standard"
              value={values.email}
              onChange={handleChange}
              error={touched.email && Boolean(errors.email)}
              helperText={touched.email && errors.email}
            />
            <TextField
              fullWidth
              id="password"
              name="password"
              label="Password"
              type="password"
              variant="standard"
              value={values.password}
              onChange={handleChange}
              error={touched.password && Boolean(errors.password)}
              helperText={touched.password && errors.password}
            />
            <Button
              variant="contained"
              type="submit"
              disabled={!isValid || isSubmitting}
            >
              Send
            </Button>
          </Form>
        )}
      </Formik>
    </div>
  );
};
