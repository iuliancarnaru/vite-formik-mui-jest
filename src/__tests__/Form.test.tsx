import 'jest-styled-components';
import {
  render,
  screen,
  waitFor,
  act,
  fireEvent,
} from '@testing-library/react';

import userEvent from '@testing-library/user-event';

import { MaterialUIForm2 } from '../Form2';

describe('MaterialUIForm', () => {
  const handleSubmit = jest.fn();

  it('should render', () => {
    const { container } = render(<MaterialUIForm2 onSubmit={handleSubmit} />);
    expect(container).toMatchSnapshot();
  });

  it('should change field values and call handleSubmit', async () => {
    render(<MaterialUIForm2 onSubmit={handleSubmit} />);

    const user = userEvent.setup();

    const emailField = screen.getByLabelText(/email/i);
    const passwordField = screen.getByLabelText(/password/i);
    const buttonEl = screen.getByRole('button', { name: /send/i });

    await user.type(emailField, 'test@test.com');
    await user.type(passwordField, 'verylongandcomplicated');

    expect(emailField).toHaveValue('test@test.com');
    expect(passwordField).toHaveValue('verylongandcomplicated');

    await user.click(buttonEl);

    await waitFor(() =>
      expect(handleSubmit).toHaveBeenCalledWith({
        email: 'test@test.com',
        password: 'verylongandcomplicated',
      })
    );
  });
});
