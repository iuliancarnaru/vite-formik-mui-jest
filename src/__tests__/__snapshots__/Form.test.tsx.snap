// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaterialUIForm should render 1`] = `
<div>
  <div
    style="max-width: 300px; padding: 20px;"
  >
    <form
      action="#"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1a1fmpi-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Send
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</div>
`;
